.packaging/rust/src/code/generate_api.rs

```diff
@@ -0,0 +1,94 @@
+use crate::{
+    code::utils::{
+        get_bindings_path, get_capi_path, get_local_enzyme_repo_path, get_local_rust_repo_path,
+    },
+    Cli,
+};
+
+use super::utils;
+use bindgen;
+use std::fs;
+
+/// This function can be used to generate Rust wrappers around Enzyme's [C API](https://github.com/wsmoses/Enzyme/blob/main/enzyme/Enzyme/CApi.h).
+//pub fn generate_bindings_with(capi_header: PathBuf, out_file: PathBuf) -> Result<(), String> {
+pub fn generate_bindings(args: Cli) -> Result<(), String> {
+    let enzyme_repo = get_local_enzyme_repo_path(args.enzyme);
+    let rust_repo = get_local_rust_repo_path(args.rust);
+
+    let capi_header = get_capi_path(enzyme_repo);
+    dbg!(&capi_header);
+
+    // tell cargo to re-run the builder if the header has changed
+    println!("cargo:rerun-if-changed={}", capi_header.display());
+    let content: String = fs::read_to_string(capi_header.clone()).unwrap();
+
+    let bindings = bindgen::Builder::default()
+        .header_contents("CApi.hpp", &content) // read it as .hpp so bindgen can ignore the class successfully
+        .clang_args(&[format!(
+            "-I{}",
+            utils::get_llvm_header_path(rust_repo).display()
+        )])
+        //.blacklist_item("CustomFunctionForward")
+        //.blacklist_item("DiffeGradientUtils")
+        .allowlist_type("CConcreteType")
+        .rustified_enum("CConcreteType")
+        .allowlist_type("CDerivativeMode")
+        .rustified_enum("CDerivativeMode")
+        .allowlist_type("CDIFFE_TYPE")
+        .rustified_enum("CDIFFE_TYPE")
+        .allowlist_type("LLVMContextRef")
+        .allowlist_type("CTypeTreeRef")
+        .allowlist_type("EnzymeTypeAnalysisRef")
+        .allowlist_function("EnzymeNewTypeTree")
+        .allowlist_function("EnzymeNewTypeTreeCT")
+        .allowlist_function("EnzymeFreeTypeTree")
+        .allowlist_function("EnzymeMergeTypeTree")
+        .allowlist_function("EnzymeTypeTreeOnlyEq")
+        .allowlist_function("EnzymeMergeTypeTree")
+        .allowlist_function("EnzymeTypeTreeShiftIndiciesEq")
+        .allowlist_function("EnzymeTypeTreeToString")
+        .allowlist_function("EnzymeTypeTreeToStringFree")
+        // Next two are for debugging / printning type information
+        .allowlist_function("EnzymeSetCLBool")
+        .allowlist_function("EnzymeSetCLInteger")
+        .allowlist_function("CreateTypeAnalysis")
+        .allowlist_function("ClearTypeAnalysis")
+        .allowlist_function("FreeTypeAnalysis")
+        .allowlist_function("CreateEnzymeLogic")
+        .allowlist_function("ClearEnzymeLogic")
+        .allowlist_function("FreeEnzymeLogic")
+        .allowlist_type("LLVMOpaqueModule")
+        .allowlist_function("EnzymeCreatePrimalAndGradient")
+        .allowlist_function("EnzymeCreateAugmentedPrimal")
```

Comment:
add the CreateForward while we're at it?

Reply:
Also could this just parse the whole header, or does it need the whitelist?

Reply:
I needed it in the past when we introduced C++ classes to the file, but now I handle them fine, so I removed it.

Without a whitelist bindgen creates bindings for everything it sees, so instead of a 300 LoC file it now creates a 9k LoC file covering all kind of LLVM functions like "LLVMIntrinsicIsOverloaded" and globals like "pub const __glibc_c99_flexarr_available: u32 = 1;" beside of the Enzyme bindings. I already use a specific crate offering nicer bindings to LLVM, so I don't need those generated ones. 

There isn't a real damage in having (and ignoring) those generated LLVM bindings and without the whitelist it should be slightly less maintenance work, so I guess that's a small win.